File: dl_library/csrc/tensor.cpp
#include "tensor.h"
#include <algorithm>
#include <cmath>
#include <random>

namespace torch_lite {

// ===== Factory =====
Tensor Tensor::zeros(const std::vector<int64_t>& shape) {
    Tensor t(shape);
    std::fill(t.data.begin(), t.data.end(), 0.0f);
    return t;
}
Tensor Tensor::ones(const std::vector<int64_t>& shape) {
    Tensor t(shape);
    std::fill(t.data.begin(), t.data.end(), 1.0f);
    return t;
}
Tensor Tensor::randn(const std::vector<int64_t>& shape, float mean, float stddev, unsigned seed) {
    Tensor t(shape);
    std::mt19937 gen(seed);
    std::normal_distribution<float> dist(mean, stddev);
    for (auto& v : t.data) v = dist(gen);
    return t;
}
Tensor Tensor::from_list(const std::vector<float>& data, const std::vector<int64_t>& shape) {
    return Tensor(data, shape);
}

// ===== In-place =====
void Tensor::fill_(float val) {
    std::fill(data.begin(), data.end(), val);
}
void Tensor::zero_() { fill_(0.0f); }
void Tensor::copy_(const Tensor& other) {
    data = other.data;
    shape = other.shape;
}

// ===== Shape =====
Tensor Tensor::reshape(const std::vector<int64_t>& new_shape) const {
    return Tensor(data, new_shape);
}
Tensor Tensor::transpose() const {
    int64_t r = rows(), c = cols();
    std::vector<float> out_data(r * c);
    for (int64_t i = 0; i < r; i++)
        for (int64_t j = 0; j < c; j++)
            out_data[j * r + i] = at(i, j);
    return Tensor(out_data, {c, r});
}

// ===== matmul =====
Tensor matmul(const Tensor& a, const Tensor& b) {
    int64_t M = a.rows(), K = a.cols();
    int64_t K2 = b.rows(), N = b.cols();
    if (K != K2) {
        throw std::runtime_error("matmul shape mismatch: (" + std::to_string(M) + "," +
            std::to_string(K) + ") @ (" + std::to_string(K2) + "," + std::to_string(N) + ")");
    }
    Tensor out = Tensor::zeros({M, N});
    for (int64_t i = 0; i < M; i++) {
        for (int64_t k = 0; k < K; k++) {
            float aik = a.at(i, k);
            if (aik == 0.0f) continue;
            for (int64_t j = 0; j < N; j++) {
                out.at(i, j) += aik * b.at(k, j);
            }
        }
    }
    return out;
}

// ===== Element-wise ops =====
Tensor add(const Tensor& a, const Tensor& b) {
    int64_t n = a.numel();
    Tensor out(a.shape);
    for (int64_t i = 0; i < n; i++) out.data[i] = a.data[i] + b.data[i];
    return out;
}
Tensor sub(const Tensor& a, const Tensor& b) {
    int64_t n = a.numel();
    Tensor out(a.shape);
    for (int64_t i = 0; i < n; i++) out.data[i] = a.data[i] - b.data[i];
    return out;
}
Tensor mul(const Tensor& a, const Tensor& b) {
    int64_t n = a.numel();
    Tensor out(a.shape);
    for (int64_t i = 0; i < n; i++) out.data[i] = a.data[i] * b.data[i];
    return out;
}
Tensor elem_div(const Tensor& a, const Tensor& b) {
    int64_t n = a.numel();
    Tensor out(a.shape);
    for (int64_t i = 0; i < n; i++) out.data[i] = a.data[i] / b.data[i];
    return out;
}
Tensor scale(const Tensor& a, float s) {
    int64_t n = a.numel();
    Tensor out(a.shape);
    for (int64_t i = 0; i < n; i++) out.data[i] = a.data[i] * s;
    return out;
}

// ===== broadcast_add: [M,N] + [N] -> [M,N] =====
Tensor broadcast_add(const Tensor& a, const Tensor& bias) {
    int64_t M = a.rows(), N = a.cols();
    Tensor out(a.shape);
    for (int64_t i = 0; i < M; i++)
        for (int64_t j = 0; j < N; j++)
            out.at(i, j) = a.at(i, j) + bias.at(0, j);
    return out;
}

// ===== Activation =====
Tensor relu(const Tensor& x) {
    Tensor out(x.shape);
    for (size_t i = 0; i < x.data.size(); i++)
        out.data[i] = x.data[i] > 0.0f ? x.data[i] : 0.0f;
    return out;
}
Tensor relu_backward(const Tensor& grad, const Tensor& input) {
    Tensor out(grad.shape);
    for (size_t i = 0; i < grad.data.size(); i++)
        out.data[i] = input.data[i] > 0.0f ? grad.data[i] : 0.0f;
    return out;
}
Tensor sigmoid(const Tensor& x) {
    Tensor out(x.shape);
    for (size_t i = 0; i < x.data.size(); i++) {
        float v = x.data[i];
        // numerically stable: clamp to avoid overflow
        if (v >= 20.0f) out.data[i] = 1.0f;
        else if (v <= -20.0f) out.data[i] = 0.0f;
        else out.data[i] = 1.0f / (1.0f + std::exp(-v));
    }
    return out;
}

// ===== Reduction =====
Tensor sum_dim0(const Tensor& x) {
    int64_t M = x.rows(), N = x.cols();
    Tensor out = Tensor::zeros({1, N});
    for (int64_t i = 0; i < M; i++)
        for (int64_t j = 0; j < N; j++)
            out.at(0, j) += x.at(i, j);
    return out;
}

// ===== Loss =====
float binary_cross_entropy(const Tensor& pred, const Tensor& target) {
    float loss = 0.0f;
    int64_t N = pred.rows();
    for (int64_t i = 0; i < N; i++) {
        float p = pred.at(i, 0);
        float y = target.at(i, 0);
        // clamp for numerical stability
        p = std::max(std::min(p, 1.0f - 1e-7f), 1e-7f);
        loss += -y * std::log(p) - (1.0f - y) * std::log(1.0f - p);
    }
    return loss / (float)N;
}

Tensor bce_sigmoid_backward(const Tensor& logits, const Tensor& target) {
    // d(BCE(sigmoid(logits), target)) / d(logits) = (sigmoid(logits) - target) / N
    Tensor out(logits.shape);
    int64_t N = logits.rows();
    for (int64_t i = 0; i < N; i++) {
        float x = logits.at(i, 0);
        float s;
        if (x >= 20.0f) s = 1.0f;
        else if (x <= -20.0f) s = 0.0f;
        else s = 1.0f / (1.0f + std::exp(-x));
        out.at(i, 0) = (s - target.at(i, 0)) / (float)N;
    }
    return out;
}

// ===== Outer product: [M] * [N] -> [M,N] =====
Tensor outer(const Tensor& a, const Tensor& b) {
    int64_t M = a.numel(), N = b.numel();
    Tensor out = Tensor::zeros({M, N});
    for (int64_t i = 0; i < M; i++)
        for (int64_t j = 0; j < N; j++)
            out.at(i, j) = a.data[i] * b.data[j];
    return out;
}

}  // namespace torch_lite

File: dl_library/torch_lite/tensor.py
"""
torch_lite.tensor — Tensor with autograd support, backed by C++ engine
"""
import numpy as np
from . import _torch_lite_core as _C


class Tensor:
    """Multi-dimensional array with automatic gradient tracking."""

    def __init__(self, data, requires_grad=False):
        if isinstance(data, _C.Tensor):
            self._c = data
        elif isinstance(data, np.ndarray):
            data = np.asarray(data, dtype=np.float32)
            self._c = _C.Tensor.from_numpy(data)
        elif isinstance(data, (list, tuple)):
            self._c = _C.Tensor.from_list(data, [len(data), 1])
        elif isinstance(data, Tensor):
            # Clone
            self._c = _C.Tensor.zeros(data._c.shape)
            self._c.copy_(data._c)
            if data.grad is not None:
                self.grad = Tensor(data.grad._c)
            requires_grad = data.requires_grad
        else:
            raise TypeError(f"Cannot create Tensor from {type(data)}")

        self.requires_grad = requires_grad
        self.grad = None
        self._backward_fn = None  # callable to compute gradients for inputs
        self._inputs = ()         # tensors used as input to create this one

    # --- Properties ---
    @property
    def shape(self): return tuple(self._c.shape)

    @property
    def ndim(self): return self._c.ndim

    @property
    def numel(self): return self._c.numel

    def numpy(self):
        return np.array(self._c.numpy())

    def __repr__(self):
        g = ", grad=True" if self.requires_grad else ""
        if self.numel <= 25:
            vals = self.numpy().flatten()
            s = ", ".join(f"{v:.4f}" for v in vals)
            return f"tensor([{s}], shape={list(self.shape)}{g})"
        return f"tensor(shape={list(self.shape)}{g})"

    # --- Operations ---

    def __matmul__(self, other):
        """C = A @ B.  Backward: dA = dC @ B.T, dB = A.T @ dC"""
        c = _C.matmul(self._c, other._c)
        out = Tensor(c, requires_grad=self.requires_grad or other.requires_grad)
        out._inputs = (self, other)

        def _backward():
            if out.grad is None:
                return
            dC = out.grad._c
            if self.requires_grad:
                g = _C.matmul(dC, other._c.t())
                _accum(self, g)
            if other.requires_grad:
                g = _C.matmul(self._c.t(), dC)
                _accum(other, g)
        out._backward_fn = _backward
        return out

    def __add__(self, other):
        if isinstance(other, (int, float)):
            c = _C.add(self._c, _C.Tensor.ones(self._c.shape))
            out = Tensor(c, requires_grad=self.requires_grad)
            out._inputs = (self,)
            def _backward():
                if out.grad is None: return
                _accum(self, out.grad._c)
            out._backward_fn = _backward
            return out
        c = _C.add(self._c, other._c)
        out = Tensor(c, requires_grad=self.requires_grad or other.requires_grad)
        out._inputs = (self, other)
        def _backward():
            if out.grad is None: return
            if self.requires_grad: _accum(self, out.grad._c)
            if other.requires_grad: _accum(other, out.grad._c)
        out._backward_fn = _backward
        return out

    def __sub__(self, other):
        c = _C.sub(self._c, other._c)
        out = Tensor(c, requires_grad=self.requires_grad or other.requires_grad)
        out._inputs = (self, other)
        def _backward():
            if out.grad is None: return
            if self.requires_grad: _accum(self, out.grad._c)
            if other.requires_grad: _accum(other, _C.scale(out.grad._c, -1.0))
        out._backward_fn = _backward
        return out

    def __mul__(self, other):
        if isinstance(other, (int, float)):
            c = _C.scale(self._c, float(other))
            out = Tensor(c, requires_grad=self.requires_grad)
            out._inputs = (self,)
            def _backward():
                if out.grad is None: return
                _accum(self, _C.scale(out.grad._c, float(other)))
            out._backward_fn = _backward
            return out
        c = _C.mul(self._c, other._c)
        out = Tensor(c, requires_grad=self.requires_grad or other.requires_grad)
        out._inputs = (self, other)
        def _backward():
            if out.grad is None: return
            if self.requires_grad: _accum(self, _C.mul(out.grad._c, other._c))
            if other.requires_grad: _accum(other, _C.mul(out.grad._c, self._c))
        out._backward_fn = _backward
        return out

    def __radd__(self, other): return self.__add__(other)
    def __rmul__(self, other): return self.__mul__(other)
    def __neg__(self):
        c = _C.scale(self._c, -1.0)
        out = Tensor(c, requires_grad=self.requires_grad)
        out._inputs = (self,)
        def _backward():
            if out.grad is None: return
            _accum(self, _C.scale(out.grad._c, -1.0))
        out._backward_fn = _backward
        return out

    def t(self):
        """Transpose: [M,N] → [N,M]"""
        c = self._c.t()
        out = Tensor(c, requires_grad=self.requires_grad)
        out._inputs = (self,)
        def _backward():
            if out.grad is None: return
            _accum(self, out.grad._c.t())
        out._backward_fn = _backward
        return out

    # --- In-place ---
    def zero_(self):
        self._c.zero_()
        return self

    def fill_(self, val):
        self._c.fill_(float(val))
        return self

    # --- Autograd ---
    def backward(self, grad=None):
        """Run reverse-mode autodiff starting from this tensor."""
        if grad is None:
            grad = Tensor(_C.Tensor.ones(self._c.shape))
        elif isinstance(grad, _C.Tensor):
            grad = Tensor(grad)
        if self.grad is None:
            self.grad = grad
        else:
            self.grad = self.grad + grad

        # Topological sort of the computation graph
        visited = set()
        topo = []

        def visit(t):
            if id(t) not in visited:
                visited.add(id(t))
                for inp in t._inputs:
                    if isinstance(inp, Tensor):
                        visit(inp)
                topo.append(t)

        visit(self)

        # Execute backward in reverse order
        for node in reversed(topo):
            if node._backward_fn is not None:
                node._backward_fn()

    def __hash__(self):
        return id(self)


def _accum(tensor, c_grad):
    """Accumulate gradient into tensor.grad"""
    if tensor.grad is None:
        tensor.grad = Tensor(c_grad, requires_grad=False)
    else:
        tensor.grad = Tensor(_C.add(tensor.grad._c, c_grad), requires_grad=False)


# --- High-level ops registered on Tensor ---
def _bias_add(tensor_input, bias):
    """Broadcast add: [M,N] + [1,N] → [M,N]"""
    c = _C.broadcast_add(tensor_input._c, bias._c)
    out = Tensor(c, requires_grad=tensor_input.requires_grad or bias.requires_grad)
    out._inputs = (tensor_input, bias)

    def _backward():
        if out.grad is None: return
        if tensor_input.requires_grad:
            _accum(tensor_input, out.grad._c)
        if bias.requires_grad:
            g = _C.sum_dim0(out.grad._c)
            _accum(bias, g)
    out._backward_fn = _backward
    return out


def _relu(tensor_input):
    """ReLU activation: max(0, x)"""
    c = _C.relu(tensor_input._c)
    out = Tensor(c, requires_grad=tensor_input.requires_grad)
    out._inputs = (tensor_input,)

    def _backward():
        if out.grad is None: return
        g = _C.relu_backward(out.grad._c, tensor_input._c)
        _accum(tensor_input, g)
    out._backward_fn = _backward
    return out


def _sigmoid(tensor_input):
    """Sigmoid activation: 1/(1+exp(-x))"""
    c = _C.sigmoid(tensor_input._c)
    out = Tensor(c, requires_grad=tensor_input.requires_grad)
    out._inputs = (tensor_input,)

    def _backward():
        if out.grad is None: return
        s = _C.sigmoid(tensor_input._c)
        g = _C.mul(out.grad._c, _C.mul(s, _C.sub(_C.Tensor.ones(s.shape), s)))
        _accum(tensor_input, g)
    out._backward_fn = _backward
    return out


def _bce_loss(logits, target):
    """Binary Cross Entropy with logits input. Returns loss scalar tensor."""
    loss_val = _C.binary_cross_entropy(_C.sigmoid(logits._c), target._c)
    loss = Tensor(_C.Tensor.from_list([loss_val], [1, 1]),
                  requires_grad=logits.requires_grad)
    loss._inputs = (logits,)

    def _backward():
        if loss.grad is None: return
        # dBCE/dlogits = (sigmoid(logits) - target) / N, then scale by loss.grad
        d = _C.bce_sigmoid_backward(logits._c, target._c)
        if abs(float(loss.grad.numpy().flatten()[0]) - 1.0) > 1e-6:
            d = _C.scale(d, float(loss.grad.numpy().flatten()[0]))
        _accum(logits, d)
    loss._backward_fn = _backward
    return loss
